<a href="https://colab.research.google.com/github/ImagingDataCommons/IDC-Tutorials/blob/master/notebooks/advanced_topics/dicom_private_tags_intro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Searching private DICOM tags using IDC BigQuery index

This notebook gives an introduction into how private tags can be used to search IDC content and select cohorts using SQL queries and IDC BigQuery index. Note that this is an advanced topic, and it might be easier for you to tackle this after completing the ["Getting started" tutorial series](https://github.com/ImagingDataCommons/IDC-Tutorials/blob/master/notebooks/getting_started/part3_exploring_cohorts.ipynb) that will introduce you to working with IDC using BigQuery index.

To be able to use this notebook, you will first need to complete the BigQuery prerequisites as discussed in another notebook here: https://github.com/ImagingDataCommons/IDC-Tutorials/blob/master/notebooks/getting_started/part1_prerequisites.ipynb.

This notebook was designed and tested in Google Colab. Please use the "Open in Colab" button at the top of the notebook to open it in Google Colab. You will need to do extra steps to make it work outside of Google Colab. If you need help with that, please open a thread in IDC Forum: https://discourse.canceridc.dev.

Author/contact: Andrey Fedorov

Initial release: Dec 2023

Assuming you completed the aforementioned prerequisites, please run the following cell after entering your GCP project ID in the cell below.

In [1]:
#@title Enter your Project ID
# initialize this variable with your Google Cloud Project ID!
my_ProjectID = "" #@param {type:"string"}

import os
os.environ["GCP_PROJECT_ID"] = my_ProjectID

from google.colab import auth
auth.authenticate_user()

## Background

Imaging equipment manufacturers, unfortunately, sometimes rely on private elements to communicate some of the acquisition details. Even more unfortunate is that in some cases, information encoded in such private attributes is absolutely essential for interpretation of those images. As an example, key acquisition parameters, such as b-values in Diffusion Weighted MRI (DWI), are stored in private elements.

What is a "private element"? This is an element that is not part of the DICOM standard, but instead is specific to a particular vendor.

The use of private elements is problematic for a number of reasons:
* users need to rely on other sources of information than the DICOM standard (typically, manufacturer DICOM conformance statements) to learn about those attributes
* the use of the private elements can change across the software versions
* private elements may be removed as part of image de-identification
* the specific tag group is not guaranteed to be fixed

Fortunately for IDC users, private DICOM elements are extracted and are searchable in BigQuery (see documentation here https://cloud.google.com/healthcare-api/docs/how-tos/dicom-bigquery-schema#private_data_elements).

In this notebook we demonstrate how private tags and corresponding data can be accessed using BigQuery on an example of the prostate DWI use case.



## Filtering images by DWI b-values

### Understanding where to find b-values

Private DICOM elements that are not sequences (not DICOM SQ VR) are extracted into the BigQuery column `OtherElements` in the IDC BigQuery index (`bigquery-public-data.idc_current.dicom_all`).

First thing to remember is that since private elements are not harmonized across vendors, you will first need to identify the subset of images that you want to analyze, identify the manufacturer and model name for the imaging equipment that produced this specific data, locate DICOM conformance statement and understand how the values of interest are encoded.

For the sake of this example, we will look into DWI images that are distributed as part of the [QIN-Prostate-Repeatability collection](http://doi.org/10.7937/K9/TCIA.2018.MR1CKGND). B-value is a key acquisition parameter for DWI.

Let's first get an idea about the manufacturer and model name for the MR images included in this collection.

In [ ]:
%%bigquery --project $my_ProjectID
SELECT
  DISTINCT Manufacturer, ManufacturerModelName
FROM
  `bigquery-public-data.idc_current.dicom_all`
WHERE
  collection_id = "qin_prostate_repeatability" AND Modality = "MR"

Query is running:   0%|          |

Downloading:   0%|          |

Manufacturer ManufacturerModelName
0  GE MEDICAL SYSTEMS      DISCOVERY MR750w
1  GE MEDICAL SYSTEMS            Signa HDxt

With a little help from ChatGPT, we located the DICOM conformance statement from GE titled ["HDxt, Hdi & Signa Vibrant Product Line DICOM CONFORMANCE STATEMENT"](https://www.gehealthcare.com/-/jssmedia/widen/2018/01/25/0204/gehealthcarecom/migrated/2018/02/19/0841/tic-resonance-gehc-dicom-conformance_signa-hdi-hdxt-15-30-vibrant-sw16_doc0878555_rev2_pdf.pdf?hash=2F2C737642001B6425A26406F04E614F&rev=-1).

This is the page that contains a rather long list of conformance statements from GE: https://www.gehealthcare.com/products/interoperability/dicom/magnetic-resonance-imaging-dicom-conformance-statements, which also includes one for Discovery 750w.

Opening the PDF documents for those statements, we can search for text string `b_value`, and locate the table that indicates that private element `(0043,1039)` is a sequence (array), which contains b-value as the first item in that sequence.

### Accessing specific private elements in the search query

Now that we know that b-values are in the first item of the `(0043,1039)` sequence, we can access its values.

In [ ]:
%%bigquery --project $my_ProjectID

 SELECT
    collection_id,
    StudyDate,
    StudyInstanceUID,
    SeriesInstanceUID,
    SeriesDescription,
    SOPInstanceUID,
    other_elements.Data[SAFE_OFFSET(0)] AS b_value
  FROM
    `bigquery-public-data.idc_current.dicom_all`,
    UNNEST(OtherElements) AS other_elements
  WHERE
    collection_id = "qin_prostate_repeatability"
    AND other_elements.Tag = "Tag_00431039"

Query is running:   0%|          |

Downloading:   0%|          |

collection_id   StudyDate  \
0      qin_prostate_repeatability  1994-06-20   
1      qin_prostate_repeatability  1994-06-20   
2      qin_prostate_repeatability  1994-07-02   
3      qin_prostate_repeatability  1994-06-20   
4      qin_prostate_repeatability  1994-07-02   
...                           ...         ...   
37949  qin_prostate_repeatability  1996-03-16   
37950  qin_prostate_repeatability  1996-03-16   
37951  qin_prostate_repeatability  1996-05-16   
37952  qin_prostate_repeatability  1996-05-02   
37953  qin_prostate_repeatability  1996-05-16   

                                        StudyInstanceUID  \
0      1.3.6.1.4.1.14519.5.2.1.3671.4754.121472087445...   
1      1.3.6.1.4.1.14519.5.2.1.3671.4754.121472087445...   
2      1.3.6.1.4.1.14519.5.2.1.3671.4754.288848219213...   
3      1.3.6.1.4.1.14519.5.2.1.3671.4754.121472087445...   
4      1.3.6.1.4.1.14519.5.2.1.3671.4754.288848219213...   
...                                                  ...   
37949  1.3.6.1.4.1.14519.5.2.1.3671.4754.288453744990...   
37950  1.3.6.1.4.1.14519.5.2.1.3671.4754.288453744990...   
37951  1.3.6.1.4.1.14519.5.2.1.3671.4754.431678793261...   
37952  1.3.6.1.4.1.14519.5.2.1.3671.4754.182983245842...   
37953  1.3.6.1.4.1.14519.5.2.1.3671.4754.431678793261...   

                                       SeriesInstanceUID  SeriesDescription  \
0      1.3.6.1.4.1.14519.5.2.1.3671.4754.304687867361...                DCE   
1      1.3.6.1.4.1.14519.5.2.1.3671.4754.304687867361...                DCE   
2      1.3.6.1.4.1.14519.5.2.1.3671.4754.768464301267...                DCE   
3      1.3.6.1.4.1.14519.5.2.1.3671.4754.304687867361...                DCE   
4      1.3.6.1.4.1.14519.5.2.1.3671.4754.609837242670...  T2 Weighted Axial   
...                                                  ...                ...   
37949  1.3.6.1.4.1.14519.5.2.1.3671.4754.334277435658...                DCE   
37950  1.3.6.1.4.1.14519.5.2.1.3671.4754.334277435658...                DCE   
37951  1.3.6.1.4.1.14519.5.2.1.3671.4754.628634179380...                DCE   
37952  1.3.6.1.4.1.14519.5.2.1.3671.4754.120594212806...                DCE   
37953  1.3.6.1.4.1.14519.5.2.1.3671.4754.628634179380...                DCE   

                                          SOPInstanceUID b_value  
0      1.3.6.1.4.1.14519.5.2.1.3671.4754.233985949889...       0  
1      1.3.6.1.4.1.14519.5.2.1.3671.4754.791413346679...       0  
2      1.3.6.1.4.1.14519.5.2.1.3671.4754.111765064689...       0  
3      1.3.6.1.4.1.14519.5.2.1.3671.4754.170721134678...       0  
4      1.3.6.1.4.1.14519.5.2.1.3671.4754.178563015192...       0  
...                                                  ...     ...  
37949  1.3.6.1.4.1.14519.5.2.1.3671.4754.230686910862...       0  
37950  1.3.6.1.4.1.14519.5.2.1.3671.4754.307508563992...       0  
37951  1.3.6.1.4.1.14519.5.2.1.3671.4754.179335954114...       0  
37952  1.3.6.1.4.1.14519.5.2.1.3671.4754.283952285382...       0  
37953  1.3.6.1.4.1.14519.5.2.1.3671.4754.304747914597...       0  

[37954 rows x 7 columns]

As you can see, we are getting a lot of images that have b-value 0. Remember, that each row in the IDC BigQuery index corresponds to a single DICOM file, which for most MR images corresponds to a single slice.

Let's next create a summary table that lists all b-values encountered for every series in our collection.

In [ ]:
%%bigquery --project $my_ProjectID

 SELECT
    SeriesInstanceUID,
    ANY_VALUE(SeriesDescription) as SeriesDescription,
    ARRAY_AGG(DISTINCT(other_elements.Data[SAFE_OFFSET(0)])) as b_values
  FROM
    `bigquery-public-data.idc_current.dicom_all`,
    UNNEST(OtherElements) AS other_elements
  WHERE
    collection_id = "qin_prostate_repeatability"
    AND Modality = "MR"
    AND other_elements.Tag = "Tag_00431039"
  GROUP BY
    SeriesInstanceUID

Query is running:   0%|          |

Downloading:   0%|          |

SeriesInstanceUID  \
0    1.3.6.1.4.1.14519.5.2.1.3671.4754.375717955574...   
1    1.3.6.1.4.1.14519.5.2.1.3671.4754.228015946741...   
2    1.3.6.1.4.1.14519.5.2.1.3671.4754.120064686826...   
3    1.3.6.1.4.1.14519.5.2.1.3671.4754.242726537614...   
4    1.3.6.1.4.1.14519.5.2.1.3671.4754.801672111149...   
..                                                 ...   
129  1.3.6.1.4.1.14519.5.2.1.3671.4754.336922202810...   
130  1.3.6.1.4.1.14519.5.2.1.3671.4754.609837242670...   
131  1.3.6.1.4.1.14519.5.2.1.3671.4754.105976129314...   
132  1.3.6.1.4.1.14519.5.2.1.3671.4754.137512301755...   
133  1.3.6.1.4.1.14519.5.2.1.3671.4754.226888855679...   

                  SeriesDescription b_values  
0                   DCE Subtraction      [0]  
1                               DCE      [0]  
2    Apparent Diffusion Coefficient   [1400]  
3                               DCE      [0]  
4                               DCE      [0]  
..                              ...      ...  
129                 DCE Subtraction      [0]  
130               T2 Weighted Axial      [0]  
131               T2 Weighted Axial      [0]  
132                             DCE      [0]  
133                             DCE      [0]  

[134 rows x 3 columns]

We have 134 MR series that have b-value initialized. Each of those series belongs to one of the following three groups:

1. b-value = [0]
2. b-value = [1400]
3. b-value = [0,1400]

We are interested in high-b DWI images, but you can also see that "Apparent Diffusion Coefficient" series has b-value of 1400, but it is not the DWI acquisition, but a parameter map derived from DWI.

When you do this kind of explorations, it is always helpful to visualize individual series so that you can confirm they are what you think they are, and also look at the DICOM tags. Let's modify the query slightly to add the URL that you can use to open each series in IDC viewer. This time we will use Python so that we can configure the output cell to hyperlink the URLs.



In [ ]:
from IPython.display import HTML

# BigQuery client is initialized with the ID of the project
# we specified in the beginning of the notebook!
bq_client = bigquery.Client(my_ProjectID)

selection_query = """
 SELECT
    SeriesInstanceUID,
    ANY_VALUE(SeriesDescription) as SeriesDescription,
    ARRAY_AGG(DISTINCT(other_elements.Data[SAFE_OFFSET(0)])) as b_values,
    ANY_VALUE(CONCAT('https://viewer.imaging.datacommons.cancer.gov/viewer/',
     StudyInstanceUID,
     '?SeriesInstanceUID=',
     SeriesInstanceUID)) as ohif_v2_url,
  FROM
    `bigquery-public-data.idc_current.dicom_all`,
    UNNEST(OtherElements) AS other_elements
  WHERE
    collection_id = "qin_prostate_repeatability"
    AND Modality = "MR"
    AND other_elements.Tag = "Tag_00431039"
  GROUP BY
    SeriesInstanceUID
"""

selection_result = bq_client.query(selection_query)
selection_df = selection_result.result().to_dataframe()

def make_clickable(val):
    return '<a href="{}" target="_blank">{}</a>'.format(val,val)

selection_df['ohif_v2_url'] = selection_df['ohif_v2_url'].apply(make_clickable)

# Display the DataFrame with clickable URLs
HTML(selection_df.to_html(escape=False))

After opening the series in the viewer, you can open "Tag Browser" to see all of the DICOM attributes, confirm your assumptions and understand the data better.

In order to filter out ADC series, we can use another - this time, standard! - DICOM attribute `ImageType`. You can see that for DWI series, the value of this attribute is `ORIGINAL\PRIMARY\OTHER`, while for ADC (since this is a derived image series), it is `DERIVED\SECONDARY\PROCESSED`.

In [ ]:
from IPython.display import HTML

# BigQuery client is initialized with the ID of the project
# we specified in the beginning of the notebook!
bq_client = bigquery.Client(my_ProjectID)

selection_query = """
 SELECT
    SeriesInstanceUID,
    ANY_VALUE(SeriesDescription) as SeriesDescription,
    ANY_VALUE(ImageType[SAFE_OFFSET(0)]) as original_or_not,
    ARRAY_AGG(DISTINCT(other_elements.Data[SAFE_OFFSET(0)])) as b_values,
    ANY_VALUE(CONCAT('https://viewer.imaging.datacommons.cancer.gov/viewer/',
     StudyInstanceUID,
     '?SeriesInstanceUID=',
     SeriesInstanceUID)) as ohif_v2_url,
  FROM
    `bigquery-public-data.idc_current.dicom_all`,
    UNNEST(OtherElements) AS other_elements
  WHERE
    collection_id = "qin_prostate_repeatability"
    AND Modality = "MR"
    AND other_elements.Tag = "Tag_00431039"
  GROUP BY
    SeriesInstanceUID
"""

selection_result = bq_client.query(selection_query)
selection_df = selection_result.result().to_dataframe()

def make_clickable(val):
    return '<a href="{}" target="_blank">{}</a>'.format(val,val)

selection_df['ohif_v2_url'] = selection_df['ohif_v2_url'].apply(make_clickable)

# Display the DataFrame with clickable URLs
HTML(selection_df.to_html(escape=False))

Now we can finally define the selection criteria for high-b DWI images. They should:
* have first element of `ImageType` set to "ORIGINAL"
* have b-value stored in the first element of private tag `(0043,1039)` set to "1400"

The query below will select series that meet these criteria, and count the slices.

In [ ]:
from IPython.display import HTML

# BigQuery client is initialized with the ID of the project
# we specified in the beginning of the notebook!
bq_client = bigquery.Client(my_ProjectID)

selection_query = """
 SELECT
    ANY_VALUE(PatientID) as PatientID,
    ANY_VALUE(StudyDate) as StudyDate,
    SeriesInstanceUID,
    ANY_VALUE(SeriesDescription) as SeriesDescription,
    ARRAY_AGG(DISTINCT(other_elements.Data[SAFE_OFFSET(0)])) as b_values,
    COUNT(DISTINCT(SOPInstanceUID)) as n_slices
  FROM
    `bigquery-public-data.idc_current.dicom_all`,
    UNNEST(OtherElements) AS other_elements
  WHERE
    collection_id = "qin_prostate_repeatability"
    AND Modality = "MR"
    AND other_elements.Tag = "Tag_00431039"
    AND ImageType[SAFE_OFFSET(0)] = "ORIGINAL"
    AND other_elements.Data[SAFE_OFFSET(0)] = "1400"
  GROUP BY
    SeriesInstanceUID
  ORDER BY
    PatientID,
    StudyDate
"""

selection_result = bq_client.query(selection_query)
selection_df = selection_result.result().to_dataframe()

# Display the DataFrame with clickable URLs
HTML(selection_df.to_html(escape=False))


### Filtering by specific b-value and downloading images

Now that we know how to identify DWI images with multiple b-values, we can filter images that correspond to the high b-value (this is often the image that is used by AI PCa detection tools).

In the query below, we get AWS URLs for all slices that have b-value of 1400. For convenience, we use `PatientID` and `SeriesInstanceUID` attributes to sort the files on the fly as we download them.

The query selects the specific series - you can remove the corresponding check from the query to select all series!

In [ ]:
from google.cloud import bigquery

# BigQuery client is initialized with the ID of the project
# we specified in the beginning of the notebook!
bq_client = bigquery.Client(my_ProjectID)

selection_query = """
WITH b_values_added AS (
 SELECT
    CONCAT("cp ",aws_url," ./",PatientID,"/",SeriesInstanceUID,"/") as aws_cp_command,
    other_elements.Data[SAFE_OFFSET(0)] AS b_value
  FROM
    `bigquery-public-data.idc_current.dicom_all`,
    UNNEST(OtherElements) AS other_elements
  WHERE
    collection_id = "qin_prostate_repeatability"
    AND other_elements.Tag = "Tag_00431039"
    AND other_elements.Data[SAFE_OFFSET(0)] = "1400"
    AND ImageType[SAFE_OFFSET(0)] = "ORIGINAL"
    # remove the next line to select all high-b images!
    AND SeriesInstanceUID = "1.3.6.1.4.1.14519.5.2.1.3671.4754.234511234806445476378897256789"
  ORDER BY
    aws_cp_command
)

SELECT
  aws_cp_command
FROM
  b_values_added
"""

selection_result = bq_client.query(selection_query)
selection_df = selection_result.result().to_dataframe()

# save to csv
selection_df.to_csv("aws_manifest.s5cmd", header=False, index=False)


Now that we have the manifest, we can [install `s5cmd`](https://github.com/peak/s5cmd#installation) and download the corresponding files.

In [ ]:
!wget https://github.com/peak/s5cmd/releases/download/v2.2.2/s5cmd_2.2.2_Linux-64bit.tar.gz
!tar -xvf ./s5cmd_2.2.2_Linux-64bit.tar.gz
!mv s5cmd /usr/local/bin
!which s5cmd

--2023-12-04 21:05:18--  https://github.com/peak/s5cmd/releases/download/v2.2.2/s5cmd_2.2.2_Linux-64bit.tar.gz
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/73909333/e095ae85-9acf-4dcc-b744-128b3311849c?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20231204%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20231204T210519Z&X-Amz-Expires=300&X-Amz-Signature=48552682d9d601ca523e8ed361abcf60145a14216a9a840c3eec5171f70767d4&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=73909333&response-content-disposition=attachment%3B%20filename%3Ds5cmd_2.2.2_Linux-64bit.tar.gz&response-content-type=application%2Foctet-stream [following]
--2023-12-04 21:05:19--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/73909333/e095ae85-9acf-4dcc-b744-

In [ ]:
!rm -rf dwi_images
!mkdir -p dwi_images
!cd dwi_images && s5cmd --no-sign-request --endpoint-url https://s3.amazonaws.com run ../aws_manifest.s5cmd

cp s3://idc-open-data/292d229b-c501-4b3d-8df7-c8615d13df20/665b7998-49de-40bb-ab2c-313afa7c9fdf.dcm PCAMPMRI-00009/1.3.6.1.4.1.14519.5.2.1.3671.4754.234511234806445476378897256789/665b7998-49de-40bb-ab2c-313afa7c9fdf.dcm
cp s3://idc-open-data/292d229b-c501-4b3d-8df7-c8615d13df20/08fb7ec2-8544-4584-a801-1c1ec16d6aa2.dcm PCAMPMRI-00009/1.3.6.1.4.1.14519.5.2.1.3671.4754.234511234806445476378897256789/08fb7ec2-8544-4584-a801-1c1ec16d6aa2.dcm
cp s3://idc-open-data/292d229b-c501-4b3d-8df7-c8615d13df20/08acac1e-b420-4185-aa00-840bd91b159f.dcm PCAMPMRI-00009/1.3.6.1.4.1.14519.5.2.1.3671.4754.234511234806445476378897256789/08acac1e-b420-4185-aa00-840bd91b159f.dcm
cp s3://idc-open-data/292d229b-c501-4b3d-8df7-c8615d13df20/3f3454e0-e03e-42b0-a816-881abcb6b0d5.dcm PCAMPMRI-00009/1.3.6.1.4.1.14519.5.2.1.3671.4754.234511234806445476378897256789/3f3454e0-e03e-42b0-a816-881abcb6b0d5.dcm
cp s3://idc-open-data/292d229b-c501-4b3d-8df7-c8615d13df20/482e1a57-0d18-4664-aced-91e34b5e10d8.dcm PCAMPMRI-00009/1

### Working with the downloaded images

In the previous section, we downloaded high-b DWI images and organized them by series.

Typically, AI tools expect input in formats such as NIfTI or NRRD, and this can be accomplished easily by converting from DICOM. For DICOM to NIfTI conversion, we recommend `dcm2niix`, which is very easy to install.

In [ ]:
%%capture
!apt-get install dcm2niix

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libyaml-cpp0.7 pigz
The following NEW packages will be installed:
  dcm2niix libyaml-cpp0.7 pigz
0 upgraded, 3 newly installed, 0 to remove and 15 not upgraded.
Need to get 417 kB of archives.
After this operation, 1,393 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 pigz amd64 2.6-1 [63.6 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/main amd64 libyaml-cpp0.7 amd64 0.7.0+dfsg-8build1 [97.7 kB]
Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 dcm2niix amd64 1.0.20211006-1build1 [256 kB]
Fetched 417 kB in 1s (555 kB/s)
Selecting previously unselected package pigz.
(Reading database ... 120882 files and directories currently installed.)
Preparing to unpack .../archives/pigz_2.6-1_amd64.deb ...
Unpacking pigz (2.6-1) ...
Selecting previously unselected package libyaml-cpp

In the next cell we convert one of the high-b DWI images into NIfTI.

In [ ]:
!rm -rf nifti_converted
!mkdir -p nifti_converted
!dcm2niix -o ./nifti_converted /content/dwi_images/PCAMPMRI-00009/1.3.6.1.4.1.14519.5.2.1.3671.4754.234511234806445476378897256789

Chris Rorden's dcm2niiX version v1.0.20211006  (JP2:OpenJPEG) GCC11.2.0 x86-64 (64-bit Linux)
Found 24 DICOM file(s)
DICOM images may be missing, expected 48 spatial locations per volume, but found 24 slices.
GE DTI directions require head first supine acquisition
Convert 24 DICOM as ./nifti_converted/1.3.6.1.4.1.14519.5.2.1.3671.4754.234511234806445476378897256789_Prostate_With_e-coil_C-_19950701142721_8 (256x256x24x1)
Conversion required 0.025703 seconds (0.025604 for core code).


Visualization of the high-b DWI image can also be done right here in the notebook using `itkWidgets`, which can read image directly from DICOM. See more details in this tutorial: https://github.com/ImagingDataCommons/IDC-Tutorials/blob/master/notebooks/getting_started/part3_exploring_cohorts.ipynb.

Installing `itkWidgets` will take some time!

In [ ]:
!pip install -q "itk-io>=5.3.0" "itkwidgets[all]>=1.0a32"

In [ ]:
import itk
from itkwidgets import view

image = itk.imread("/content/dwi_images/PCAMPMRI-00009/1.3.6.1.4.1.14519.5.2.1.3671.4754.234511234806445476378897256789")

viewer = view(image, ui_collapsed=False)


viewer.set_background_color([0.1,0.1,0.1])
viewer.set_image_color_map("Grayscale")
viewer.set_image_color_range([-1727,273])
viewer.set_label_image_blend(0.75)
viewer.set_view_mode('ZPlane')

## Looking beyond QIN-Prostate-Repeatability

The query below will summarize all of the collections that have the notorious GE-specific b-values tag. As you can see, some of the values appear to follow different conventions, such as 1000000600, which cannot be realistic.

In [3]:
%%bigquery --project $my_ProjectID

WITH
  b_value_added AS (
  SELECT
    collection_id,
    Manufacturer,
    ManufacturerModelName,
    ARRAY_TO_STRING(SoftwareVersions,"/"),
    PatientID,
    StudyDate,
    StudyInstanceUID,
    SeriesInstanceUID,
    SeriesDescription,
    SOPInstanceUID,
    other_elements.Data[SAFE_OFFSET(0)] AS b_value
  FROM
    `bigquery-public-data.idc_current.dicom_all`,
    UNNEST(OtherElements) AS other_elements
  WHERE
    other_elements.Tag = "Tag_00431039" )
SELECT
  collection_id,
  array_to_string(ARRAY_AGG(DISTINCT(b_value)),",") as b_values_encountered,
  ARRAY_AGG(DISTINCT(CONCAT(SAFE_CAST(Manufacturer AS STRING)," ",ManufacturerModelName)) IGNORE NULLS) as Manufacturer_and_ModelName
FROM
  b_value_added
WHERE
  b_value <> ""
GROUP BY
  collection_id

Query is running:   0%|          |

Downloading:   0%|          |

collection_id  \
0                       tcga_prad   
1           breast_mri_nact_pilot   
2                       tcga_lihc   
3                         cmb_mml   
4                       tcga_brca   
5                      acrin_6698   
6                       tcga_kirc   
7                       tcga_blca   
8                       tcga_cesc   
9           pseudo_phi_dicom_data   
10         prostate_mri_us_biopsy   
11            acrin_nsclc_fdg_pet   
12                          ispy2   
13                        cmb_lca   
14                       cptac_cm   
15                    icdc_glioma   
16     qin_prostate_repeatability   
17                          ispy1   
18                      tcga_ucec   
19                covid_19_ny_sbu   
20                     cptac_ucec   
21                      tcga_sarc   
22                      cptac_sar   
23                      cptac_pda   
24                     cptac_luad   
25                      tcga_kirp   
26  acrin_contralateral_breast_mr   
27                    cptac_ccrcc   
28                        cmb_crc   

                                 b_values_encountered  \
0                                               0,600   
1                                           0,450,600   
2                                               0,500   
3                                                   0   
4                                              0,1000   
5      800,1000000800,100,0,1000000600,1000000100,600   
6                                          500,1000,0   
7                                               0,800   
8                                           800,0,100   
9                                                   0   
10                                                  0   
11                                        1200,0,1000   
12                                     1000,100,0,500   
13                                             0,1000   
14                                             1000,0   
15                                  1000001000,0,1000   
16                                             1400,0   
17                                              0,600   
18                                                  0   
19                            0,1000000050,1000000600   
20                                       0,1000000700   
21                                                  0   
22                                                  0   
23  600,1000001200,10,1000000900,1000000500,0,1000...   
24                                             0,1000   
25  0,1000000800,1000000500,1000,1000001000,100000...   
26                                                  0   
27                            0,1000000020,1000000800   
28      0,1000000400,1000000800,1000000050,1000000100   

                           Manufacturer_and_ModelName  
0   [GE MEDICAL SYSTEMS GENESIS_SIGNA, GE MEDICAL ...  
1                  [GE MEDICAL SYSTEMS GENESIS_SIGNA]  
2   [GE MEDICAL SYSTEMS GENESIS_SIGNA, GE MEDICAL ...  
3   [GE MEDICAL SYSTEMS Signa HDxt, GE MEDICAL SYS...  
4   [GE MEDICAL SYSTEMS SIGNA HDx, GE MEDICAL SYST...  
5   [GE MEDICAL SYSTEMS Signa HDxt, GE MEDICAL SYS...  
6   [GE MEDICAL SYSTEMS SIGNA HDx, GE MEDICAL SYST...  
7   [GE MEDICAL SYSTEMS Signa HDxt, GE MEDICAL SYS...  
8   [GE MEDICAL SYSTEMS SIGNA EXCITE, GE MEDICAL S...  
9                  [GE MEDICAL SYSTEMS GENESIS_SIGNA]  
10  [GE MEDICAL SYSTEMS DISCOVERY MR750w, GE MEDIC...  
11  [GE MEDICAL SYSTEMS SIGNA EXCITE, GE MEDICAL S...  
12  [GE MEDICAL SYSTEMS DISCOVERY MR750w, GE MEDIC...  
13                 [GE MEDICAL SYSTEMS Optima MR450w]  
14                    [GE MEDICAL SYSTEMS Signa HDxt]  
15  [GE MEDICAL SYSTEMS GENESIS_SIGNA, GE MEDICAL ...  
16  [GE MEDICAL SYSTEMS DISCOVERY MR750w, GE MEDIC...  
17  [GE MEDICAL SYSTEMS GENESIS_SIGNA, GE MEDICAL ...  
18                    [GE MEDICAL SYSTEMS Signa HDxt]  
19                [GE MEDICAL SYSTEMS SIGNA Explorer]  
20  [GE MEDICAL SYSTEM

# Next steps

Share your feedback or ask questions about this tutorial in IDC Forum: https://discourse.canceridc.dev.